nemo/collections/asr/parts/k2/maploss.py

```
@@ -0,0 +1,269 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Union
+
+import k2
+import torch
+
+from nemo.collections.asr.parts.k2.autograd import sparse_abs
+from nemo.collections.asr.parts.k2.utils import (
+    create_sparse_wrapped,
+    create_supervision,
+    get_tot_objf_and_finite_mask,
+    load_graph,
+    make_blank_first,
+    prep_padded_densefsavec,
+    shift_labels_inpl,
+)
+from nemo.utils import logging
+
+
+class MAPLoss(torch.nn.Module):
+    """
+    Maximum a Posteriori Probability criterion.
+    It is implemented as Lattice-Free Maximum Mutual Information (LF-MMI) 
+    and LF-boosted-MMI (LF-bMMI) losses.
+    
+    Based on https://github.com/k2-fsa/snowfall/blob/master/snowfall/objectives/mmi.py
+    """
+
+    def __init__(
+        self,
+        num_classes: int,
+        blank: int,
+        reduction: str = "mean",
+        topo_type: str = "default",
+        topo_with_selfloops: bool = True,
+        loss_type: str = "mmi",
+        token_lm: Optional[Union[k2.Fsa, str]] = None,
+        intersect_pruned: bool = False,
+        boost_coeff: float = 0.0,
+        **kwargs,
+    ):
+        super().__init__()
+        self.num_classes = num_classes
+        self.blank = blank
+        self.reduction = reduction
+        self.loss_type = loss_type
+        self.boost_coeff = boost_coeff
+        self.intersect_calc_scores = (
+            self._intersect_calc_scores_mmi_pruned if intersect_pruned else self._intersect_calc_scores_mmi_exact
+        )
+        self.topo_type = topo_type
+        self.topo_with_selfloops = topo_with_selfloops
+        self.pad_fsavec = topo_type == "compact"
+        self.graph_compiler = None
+        if token_lm is None:
+            logging.warning(
```

Isnt this an invalid state? Maybe error out rather than warn?